File: Week2/수업용/그래프_탐색/네트워크/Solution_Network_BFS.py
from collections import deque # 효율적인 큐(Queue) 사용을 위해 deque 라이브러리를 가져옵니다.
import re # 문자열에서 숫자만 추출하기 위해 정규표현식 라이브러리를 가져옵니다.

def solution(n, computers):
    """
    네트워크의 개수를 구하는 함수
    :param n: 컴퓨터의 개수 (정수)
    :param computers: 연결 정보를 담은 2차원 리스트 (인접 행렬)
    :return: 연결된 네트워크의 총 개수
    """
    answer = 0 # 네트워크의 개수를 저장할 변수
    visited = [False] * n # 각 컴퓨터를 방문했는지 확인하는 리스트 (처음엔 모두 False)

    # 0번 컴퓨터부터 n-1번 컴퓨터까지 하나씩 확인합니다.
    for i in range(n):
        # 만약 아직 방문하지 않은 컴퓨터를 발견한다면
        if not visited[i]:
            # 새로운 네트워크 탐색을 위해 큐(Queue)를 만들고 현재 컴퓨터를 넣습니다.
            queue = deque([i])
            visited[i] = True # 현재 컴퓨터를 방문했다고 표시합니다.

            # 큐가 빌 때까지(연결된 모든 컴퓨터를 찾을 때까지) 반복합니다.
            while queue:
                v = queue.popleft() # 큐에서 가장 먼저 들어온 컴퓨터를 하나 꺼냅니다.
                
                # 현재 꺼낸 컴퓨터(v)와 다른 컴퓨터들(neighbour)의 연결 상태를 확인합니다.
                for neighbour in range(n):
                    # v와 neighbour가 연결되어 있고(1), neighbour를 아직 방문하지 않았다면
                    if computers[v][neighbour] == 1 and not visited[neighbour]:
                        visited[neighbour] = True # neighbour 컴퓨터를 방문 표시합니다.
                        queue.append(neighbour) # 다음 탐색을 위해 큐에 추가합니다.
            
            # BFS 탐색이 한 번 끝나면 하나의 네트워크를 다 찾은 것이므로 개수를 1 증가시킵니다.
            answer += 1
            
    return answer # 최종적으로 계산된 네트워크 개수를 반환합니다.


if __name__ == "__main__":
    # 사용자로부터 컴퓨터의 개수를 입력받습니다.
    n = int(input("컴퓨터의 개수를 입력하세요: "))
    print("인접 행렬을 입력하세요: ")

    # 입력을 처리하는 부분입니다. (대괄호나 쉼표가 포함되어도 숫자만 골라냅니다.)
    raw_text = ""
    # n * n개의 숫자가 채워질 때까지 입력을 계속 받습니다.
    while len(re.findall(r'\d+', raw_text)) < n * n:
        raw_text += " " + input()
        
    # 추출한 숫자들을 정수 리스트로 변환합니다.
    all_nums = list(map(int, re.findall(r'\d+', raw_text)))
    
    # 1차원 리스트를 n x n 형태의 2차원 리스트로 다시 구성합니다.
    computers = [all_nums[i * n: (i + 1) * n] for i in range(n)]

    # 결과를 출력합니다.
    print(f"결과: {solution(n, computers)}")

File: Week2/수업용/그래프_탐색/네트워크/Solution_Network_BFS.java
package Week2.수업용.그래프_탐색.네트워크;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

/**
 * Solution_Network_BFS 클래스: 너비 우선 탐색(BFS)을 이용해 네트워크 개수를 구하는 클래스
 */
public class Solution_Network_BFS {
    /**
     * solution 함수: 전체 네트워크 개수를 계산합니다.
     * @param n: 컴퓨터의 개수
     * @param computers: 컴퓨터 간의 연결 여부를 나타내는 2차원 배열 (1: 연결됨, 0: 연결 안 됨)
     * @return: 연결된 독립적인 네트워크의 총 개수
     */
    public int solution(int n, int[][] computers) {
        int answer = 0; // 네트워크 개수를 세기 위한 변수
        boolean[] visited = new boolean[n]; // 각 컴퓨터를 방문했는지 체크하는 배열

        // 모든 컴퓨터(0번부터 n-1번까지)를 하나씩 살펴봅니다.
        for (int i = 0; i < n; i++) {
            // 만약 아직 방문하지 않은 컴퓨터라면, 새로운 네트워크의 시작점으로 봅니다.
            if (!visited[i]) {
                // BFS 탐색을 위해 큐(Queue)를 생성합니다.
                Queue<Integer> queue = new LinkedList<>();
                queue.offer(i); // 시작 컴퓨터를 큐에 넣습니다.
                visited[i] = true; // 시작 컴퓨터를 방문했다고 표시합니다.

                // 큐에 탐색할 컴퓨터가 남아있는 동안 계속 반복합니다.
                while (!queue.isEmpty()) {
                    int v = queue.poll(); // 큐에서 컴퓨터 한 대를 꺼냅니다.
                    
                    // 현재 꺼낸 컴퓨터(v)와 연결된 다른 모든 컴퓨터를 확인합니다.
                    for (int neighbour = 0; neighbour < n; neighbour++) {
                        // 만약 두 컴퓨터가 연결되어 있고(1), 
                        // 상대 컴퓨터를 아직 방문하지 않았다면
                        if (computers[v][neighbour] == 1 && !visited[neighbour]) {
                            visited[neighbour] = true; // 상대 컴퓨터를 방문 표시합니다.
                            queue.offer(neighbour); // 다음 탐색을 위해 큐에 넣습니다.
                        }
                    }
                }
                // 하나의 연결된 덩어리(네트워크)를 다 찾았으므로 개수를 1 증가시킵니다.
                answer++;
            }
        }
        return answer; // 찾은 네트워크의 총 개수를 반환합니다.
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        // 숫자 이외의 문자(예: [, ], , 등)를 입력받을 때 무시하기 위한 설정입니다.
        sc.useDelimiter("[^0-9]+");

        System.out.print("컴퓨터의 개수를 입력하세요: ");
        int n = sc.nextInt(); // 컴퓨터의 개수를 입력받습니다.
        int[][] computers = new int[n][n]; // n x n 크기의 연결 정보 배열을 만듭니다.

        System.out.println("인접 행렬을 입력하세요: ");
        // 2차원 배열의 모든 칸을 돌며 연결 정보를 입력받습니다.
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                computers[i][j] = sc.nextInt();
            }
        }

        // solution 함수를 호출하여 결과를 구하고 출력합니다.
        Solution_Network_BFS sol = new Solution_Network_BFS();
        System.out.println("결과: " + sol.solution(n, computers));
    }
}
